book/src/passive-stake-delegation-and-rewards.md

```diff
@@ -0,0 +1,149 @@
+rewards, and therefore the number of stakes is expected to be large compared to
+the number of validators in a long running cluster.
+
+## Proposed changes to the current design.
+
+The general idea is that instead of the staker, the validator will own the
+VoteState program. In this proposal the VoteState program is there to track
+validator votes, count validator generated credits and to provide any
+additional validator specific state.  The VoteState program is not aware of any
+stakes delegated to it, and has no staking weight.
+
+The rewards generated are proportional to the amount of lamports staked.  In
+this proposal stake state is stored as part of the RewardsState program. This
+program is owned by the staker only.  Lamports stored in this program are the
+stake.  Unlike the current design, this program contains a new field to indicate
+which VoteState program the stake is delegated to.
+
+### New VoteState
+
+VoteState is the current state of all the votes the **delegate** has submitted
+to the bank.  VoteState contains the following state information:
+
+* votes - The submitted votes data structure.
+
+* credits - The total number of rewards this vote program generated over its
+lifetime.
+
+* root\_slot - The last slot to reach the full lockout commitment necessary for
+rewards.
+
+* commission - The commission taken by this VoteState for any rewards claimed by
+staker's RewardsState accounts.
+
+* lamports - The accumulated lamports from the commission.  These do not count as
+stakes.
+
+* `authorized_voter_id` - Only this identity is authorized to submit votes.
+
+### New RewardsState
+
+RewardsState is the current delegation preference of the **staker**. RewardsState
+contains the following state information:
+
+* lamports - The staked lamports.
+
+* `delegate` - The pubkey of the VoteState instance the lamports are
+delegated to.
+
+* `claimed_credits` - The total credits claimed over the lifetime of the
+program.
+
+
+### Claiming Rewards
+
+The VoteState program and the RewardsState programs maintain a lifetime counter
+of total rewards generated and claimed.  Therefore an explicit `Clear`
+instruction is not necessary.  When claiming rewards, the total lamports
+deposited into the RewardsState and as validator commission is proportional to
+`VoteState::credits - RewadsState::claimed_credits`.
+
+### RewardsInstruction::Initialize
+
+* `account[0]` - Out Param - The RewardsState instance.  
+  `RewardsState::claimed_credits` is initialized to `VoteState::credits`.  
+  `RewardsState::delegate` is initialized to `account[1]`
+
+* `account[1]` - In Param - The VoteState instance.
+
+### RewardsInstruction::RedeemVoteCredits
+
+
+* `account[0]` - Out Param - The RewardsState instance.  
+
+* `account[1]` - In Param - The VoteState instance, must be the same as
+`RewardsState::delegate`
+
+
+Reward is payed out for the difference between `VoteState::credits` to
```

Comment:
Payed by whom?

Reply:
By the bank.  Generated out of thin air.

Reply:
Alternative is to have a single synchronization point for the pool that is RW at stake creation.  But transaction fees deposit into that account anyways, so we need to special case it.

Reply:
Current Rewards program is more elegant. Without a program, any change to the fees algorithm would require every validator in the cluster will have to update its code.

Reply:
@garious How does it collect the transaction fees?

Reply:
What is "it"?

Reply:
How do transaction fees get deposited into the mining pool?

Reply:
There's nothing preventing the leader from depositing a portion of the fees into the Rewards account.

Reply:
@garious with a transaction?  But if the leader doesn't do it, do the validators fail that block?

Reply:
Just hard-coded into the protocol, same as the leader's portion. Yes, validators should fail the block if portions are inconsistent.

Reply:
@garious then it is the same elegance problem as keeping the mining pool completely virtual. The remaining amount can be calculated by omission by all the validators.  If any behavior changes, then the same hard forking upgrade is required anyways.

It's easy enough to add a single account to store the total, and use it as a synchronous creation point for new stakes. But because the fee deposits require the "out of program" behavior it doesn't seem like a worthwhile gain over just keeping it virtual.

Reply:
But, its kind of nice not special case the Bank to have extra tokens appear.  The pipeline can deposit the fees at the end of each block, so the next set of RewardsState programs can claim them.

Reply:
Any way I can convince you to abandon your redefinition of the Rewards program? Have you read through this in detail yet?  https://github.com/solana-labs/solana/pull/3212.  I don't think we'll be able to take action until we converge.  I don't feel strongly that the Rewards logic needs to be in an on-chain program, but if you can avoid redefining it, we can move forward by having `solana_runtime` depend on `solana_rewards_api` and calling a function that would have otherwise been called by sending the Rewards account a `RedeemVoteCredits` instruction. Then we can have a separate discussion about whether it's "better" to invoke that function with a transaction or with a hard-coded function call in the bank.  Again, I don't feel strongly about that. I do, however, feel strongly against your attempt to morph Rewards program into what my proposal is calling the Stake program. I don't see any good reason to do that, and it makes this discussion unnecessarily confusing. In your last comment above, for example, I have no idea what you're talking about when you say "RewardsState programs can claim them."

Reply:
So just s/RewardsState/StakeState/g ?

Reply:
Yes. And I think you're proposing deleting RewardsState and moving that logic into Bank.

Reply:
Sort of.  I am proposing the bank allow the StakeState program to bypass the token balance check.  It can work without it, but claiming the reward will have a synchronization point for the entire network.

Reply:
Hence https://github.com/solana-labs/solana/blob/master/programs/rewards/src/lib.rs#L14

Reply:
@garious updated.  Mining Pool is run as a program, and the end of the block the bank executes a virtual transaction to collect the fees for the leader, and the mining pool.

Reply:
What's the difference between the new MiningPool and old Rewards accounts? Just which program owns it?

Reply:
@garious yes.

the Stake is RW to modify the claimed_credits counter, and the MiningPool is RW to spend the pool tokens.